google/cloud/aiplatform/prediction/model_server.py

```diff
@@ -0,0 +1,190 @@
+        ):
+            raise ValueError(
+                "Both of the environment variables AIP_HEALTH_ROUTE and "
+                "AIP_PREDICT_ROUTE need to be specified."
+            )
+        self.http_port = int(os.environ.get("AIP_HTTP_PORT"))
+        self.health_route = os.environ.get("AIP_HEALTH_ROUTE")
+        self.predict_route = os.environ.get("AIP_PREDICT_ROUTE")
+
+        self.app = FastAPI()
+        self.app.add_api_route(
+            path=self.health_route,
+            endpoint=self.health,
+            methods=["GET"],
+        )
+        self.app.add_api_route(
+            path=self.predict_route,
+            endpoint=self.predict,
+            methods=["POST"],
+        )
+
+    async def __call__(self, scope, receive, send):
+        await self.app(scope, receive, send)
+
+    def _init_logging(self):
+        """Initializes the logging config."""
+        logging.basicConfig(
+            format="%(asctime)s: %(message)s",
+            datefmt="%m/%d/%Y %I:%M:%S %p",
+            level=logging.INFO,
+        )
+
+    def health(self):
+        """Executes a health check."""
+        return {}
+
+    async def predict(self, request: Request) -> Response:
+        """Executes a prediction.
+
+        Args:
+            request (Request):
+                Required. The prediction request.
+
+        Returns:
+            The response containing prediction results.
+        """
+        try:
+            return await self.handler.handle(request)
+        except HTTPException:
+            # Raises exception if it's a HTTPException.
+            raise
+        except Exception as exception:
+            error_message = "An exception {} occurred. Arguments: {}.".format(
+                type(exception).__name__, exception.args
+            )
+            logging.info(
+                "{}\\nTraceback: {}".format(error_message, traceback.format_exc())
+            )
+
+            # Converts all other exceptions to HTTPException.
+            raise HTTPException(status_code=500, detail=error_message)
+
+
+def set_number_of_workers_from_env() -> None:
+    """Sets the number of model server workers used by Uvicorn in the environment variable.
+
+    The number of model server workers will be set as WEB_CONCURRENCY in the environment
+    variables.
+    The default number of model server workers is the number of cores.
+    The following environment variables will adjust the number of workers:
+        WEB_CONCURRENCY:
+            The number of the workers. This will overwrite the number calculated by
+            the other variables, min(WORKERS_PER_CORE * number_of_cores, MAX_WORKERS).
+        WORKERS_PER_CORE:
+            The number of the workers per core. The default is 1.
+        MAX_WORKERS:
+            The maximum number of workers can be used given the value of WORKERS_PER_CORE
+            and the number of cores.
+    """
+    workers_per_core_str = os.getenv("WORKERS_PER_CORE", "1")
```

Comment:
nit: consider renaming all these env variables to have a "VERTEX_CPR_" prefix or something similar. It groups all the env variables, making debugging easier, and lets users know what is setting and using these env variables.

Reply:
Since the model server (this file) will run in the built images, users who build the images should know they are using CPR. I prefer to leave it as is (without the prefix).

Reply:
Also, `WEB_CONCURRENCY` is the parameter to change the number of workers for uvicorn and it's not set from our code.